docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -617,9 +588,58 @@ The fact that the DDL procedure in TiDB is mature helps to achieve some features
 - Updating schema version guarantees all active transactions are based on the same version of placement rules
 
+### Building placement rules
+
+There needs a way to map the placement rules in SQL to PD placement rule configuration. Most of the fields are discussed above, so this part focuses on `group_id`, `id`, `start_key` and `end_key`.
+
+`group_id` is used to identify the source of the placement rules, so `group_id` is `tidb`.
+
+`ALTER PLACEMENT POLICY` and `DROP PLACEMENT POLICY` need to find the rules of a specified object efficiently. It can be achieved by encoding the object ID in `id`.
+
+However, an object (database, table, partition) may have multiple rules for a single role. For example:
+
+```sql
+ALTER TABLE t
+	PLACEMENT CONSTRAINTS="{+zone=bj:2,+zone=sh:1}" ROLE=voter;
+```
+
+It needs 2 placement rules for `voter` in the PD placement rule configuration, because each rule can only specify one `count`. To make `id` unique, a unique identifier must be appended to `id`. DDL job ID plus an index in the job is a good choice.
+
+Take the case above for example, assuming the table ID of `t` is 100, the ID of the DDL job executing this statement is 200, then `id` of the placement rules are `100-200-1` and `100-200-2`.
+
+The prefix of `id` is in such a format:
+
+* Database: database id
+* Table: table id
+* Partition: partition id
+* Unpartitioned index: the concatenation of table id and index id, e.g. `100_1`
+* Partitioned index: the concatenation of partition id and index id
+
+To query all the placement rules for one object, PD looks for all the `id` with a specific prefix.
+
+As all placement rules are mapped to PD placement rule configurations, `start_key` and `end_key` must be generated for each object. However, databases and partitioned tables have no key ranges, so the only way is to generate a key range with no actual records.
+
+As database IDs are all globally unique, it's fine to replace table ID with database ID in the key range. For example, assuming the database ID is 100, then the string format of its key range is:
+
+- `start_key`: `t_{database_id}_`
+- `end_key`: `t_{database_id+1}_`
+
+It's same for partitioned tables.
+
+
 ### Rule priorities
 
-When several rules are defined for one record, the most granular rule is chosen for this record. More specifically, the rule priority is: index > partition > table > database > default.
+Tables only inherit rules from databases when they are created, and the value is saved in the meta data. Thus, the rules of priorities are simplified from an earlier version of this proposal (and are more inline with how character sets are inherited).
+
+The only rules are that indexes and partitions inherit the rules of tables. Partitions can explicitly overwrite the placement policy, but indexes currently do not allow placement policy to be defined (this simplification was made intentionally since there is not a clear use case until global secondary indexes are introduced).
```

Comment:
How about the following steps:
1. a placement rule is added on table `t`
2. a placement rule is added on partition `p0` of `t`
3. the placement rule is altered on table `t`
Will `p0` use this modified placement rule?

Reply:
Yes it will. See subheadings `Partitioned Tables` and `Optimization: Latest data on SSD` for examples.